### DISCUSSION TITLE
pyuvm vs uvm-python

### DISCUSSION THREAD

**Discussion by timothyscherer** [pyuvm/pyuvm#181] [General]
I have spent the last few weeks learning pyuvm and getting set up with my first project. As the project grows and I dig deeper into learning UVM itself, I notice different elements of UVM that are not implemented in pyuvm. I did some research and came across [uvm-python](https://github.com/tpoikela/uvm-python), which seems to be a more "direct" (for lack of a better word) implementation of the UVM in Python.

After just browsing through some of the source files, I think I prefer the architectural approach, which follows the UVM source architecture very closely. It also seems to implement a lot more of the low-level details and overall looks more like UVM.

So I am curious to know, why are these two projects independent of each other? They both implement the same standard in the same language based on cocotb, with two different style approaches. Based on commit history, it seems that pyuvm was created first, but was very minimal until after the "boom" of uvm-python. uvm-python seems to be going a little stale now, whereas pyuvm seems to be gaining traction. What are your thoughts about this?

**Comment by raysalemi** [comment]
As I understand it, **uvm-python** was created using a script that read the SystemVerilog and generated Python. As a result, it contains  hacks that were necessary in SystemVerilog such as the `_imp` classes and doesn't take advantage of features in Python such as the multiple-inheritance that makes `_imp` classes unnecessary.  

I'm not sure if uvm-python works, as I've never used it.  Weirdly, I hadn't heard of it until after I wrote **pyuvm**.

**Comment by timothyscherer** [comment]
I think there is at least a lot to learn from that repo. For example, I understand that the choice was made for pyuvm to use Python's logging utilities, but uvm-python did a great job of implementing the UVM-defined logging system, and it would certainly feel more confortable (amongst other things) to someone switching from SV-UVM to pyuvm, which is where a lot of pyuvm's growth will come from.

I hope pyuvm can become a mixture of both of these repositories at some point, so that it is more SV-UVM feature-rich while also being more Pythonic and eliminating all of the SystemVerilog "workarounds" basked into UVM.

What are your thoughts on adopting the structure of uvm-python (organizing files similar to the SV-UVM source)? Personally, I am a fan of that approach since it is easier to compare to the original source. Organization based on the class reference may get confusing, especially when it comes time to upgrade from UVM 1.2 to 2.0 (?).

**Reply by raysalemi** [reply]
I am not a fan of that approach since the generated code is terrible and there is no need to recreate well-tested items from the Python ecosystem.
